### PR TITLE
Add a possibility to use a formatter on a warning panel. Applied it for a scaffolder template

### PR DESCRIPTION
**Pull request by acierto**
## Hey, I just made a Pull Request!

I would like to give a better user experience to the users to deal with error messages. 
Along with the error message, I'd like to point the user to the documentation page to read how to tackle it, see the troubleshooting guide (run deck, etc.)

At the moment the WarningPanel/ErrorPanel is limited and allows to show the text only.
The proposal is to allow to enable markdown capabilities to achieve that purpose. 

The example how it can look

Screenshot 2023-10-04 at 12 51 40

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
I wonder if there's any harm in just applying this for all the `ErrorPanel` components? @freben @Rugvip?

File: packages/core-components/src/components/ErrorPanel/ErrorPanel.test.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    title: 'Mock title',
+    message: 'Some more info',
+  };
+
+  it('renders a title formatted by markdown', async () => {
+    await renderInTestApp(
+      <WarningPanel
+        {...propsErrorMessage}
+        formatTitle="markdown"
+        title="Step has been failed. [Help](https://commonmark.org/help)"
```
Comment (freben) [inline packages/core-components/src/components/ErrorPanel/ErrorPanel.test.tsx:35]:
```suggestion
        title="Step has failed. [Help](https://commonmark.org/help)"
```

Comment (freben) [reply packages/core-components/src/components/ErrorPanel/ErrorPanel.test.tsx:35]: same for all other places below

**Review by freben — COMMENTED** [review]
yeah in all honesty this could be the default for just the plain title, if we wanted

**Comment by acierto** [comment]
How are we going to proceed? Applying this change for everything or just for a this case?

**Comment by freben** [comment]
It's unfortunate to do this at all on the title, in my humble opinion. Titles aren't a good place for actions and links, and the formatting needs here really throw a wrench in things, just look at your `p` styling needs - it's unfortunate complexity. And the developer needs to know about the secret string "markdown" and what it means. For example accepting a `JSX.Element` as title would probably be superior and give more freedom without affecting styling.

Another idea: You can put anything inside accordions. https://v4.mui.com/components/accordion/#secondary-heading-and-columns How about perhaps adding an `actions` prop that lets you add buttons similar to what's in that example, and thereby having proper support for any type of action?

**Comment by lolsem** [comment]
@freben I'm just trying to think a long, how about adding a button in the alert instead of a link? Having either a link or action button can save the user another click. https://v4.mui.com/components/alert/ (see example with the undo action button)

For MUI alert components they offer an action button on the right side of the alert.

**Comment by freben** [comment]
> For MUI alert components they offer an action button on the right side of the alert.

You'd really want those actions aligned to the side then (and probably on the left side in RTL locales) and in a separate div that doesn't line break with the text. This just isn't suitable, I think, in the first place to mingle with the title property at all.

**Comment by freben** [comment]
> @freben Not sure how you envision to use `JSX.Element`. Template custom action is failed or by rejected promise or by thrown error. And now we send there a text message. Do you propose to do `throw new Error(<div>Error<div/>)`?

Yeah honestly I'm not sure what the contract could be. It's a bit unfortunate that it's communicated via errors, but that's what we have to work with I guess.

> About an alternative idea, how do you see the contract from custom template? Will it look like
> 
> ```
> 
> throw new Error({
>      message: 'Something went wrong',
>      buttons: {
>          label: 'Info',
>          onClick: () => window.open("https://www.w3schools.com")
>      }
> })
> ```
> 
> And then inside the Warning/ErrorPanel add buttons if they are described in error?

Yeah I think if we want to go farther than just links in the title itself is some form of scaffolder-custom Error subclass that's very specific and has more properties than the default one. Then the error panel needs to handle that error specifically and do what makes sense. This is what we do [in `ResponseErrorPanel`](https://github.com/backstage/backstage/blob/b017c80551c79368e226ed50232504413ec83045/packages/core-components/src/components/ResponseErrorPanel/ResponseErrorPanel.tsx#L55) for displaying the serialized rich backend errors.

I have to fall back to asking @benjdlambert, how would you prefer to do something like this if you got to choose? My very first take was, what if we always rendered the regular title as `MarkdownContent` no matter what. But maybe that'll just lead to further styling issues where we least expect it ...

**Comment by acierto** [comment]
@freben We discussed this PR with @benjdlambert and @Rugvip last Friday on a call. 
We came to a conclusion to have it as is would be fine, only to rename the property name from `formatTitle` to `titleFormat`.
And keep using it only for scaffolder, not the entire application.

Resolved the merge conflicts and renamed the property.

File: .changeset/gentle-elephants-look.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/core-components': patch
+'@backstage/plugin-scaffolder': patch
```
Comment (benjdlambert) [inline .changeset/gentle-elephants-look.md:3]: `minor`

**Review by benjdlambert — COMMENTED** [review]
Just one last thing here./

File: .changeset/gentle-elephants-look.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/core-components': minor
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply .changeset/gentle-elephants-look.md:2]: @acierto just this little thing left I think?

[the remark this answers is not in the document]
Comment (acierto) [reply .changeset/gentle-elephants-look.md:2]: ah, missed it, let me change it

**Review by benjdlambert — APPROVED** [review]

**Pull request merged**